src/input_common/gcadapter/gc_adapter.cpp

```diff
@@ -283,7 +285,88 @@ void Adapter::Reset() {
     }
 }
 
-bool Adapter::DeviceConnected(std::size_t port) {
+std::vector<Common::ParamPackage> Adapter::GetInputDevices() const {
+    std::vector<Common::ParamPackage> devices;
+    for (std::size_t port = 0; port < state.size(); ++port) {
+        if (DeviceConnected(port)) {
+            std::string name = fmt::format("{} {}", "Gamecube Controller", port);
+            devices.emplace_back(Common::ParamPackage{
+                {"class", "gcpad"},
+                {"display", std::move(name)},
+                {"port", std::to_string(port)},
+            });
+        }
+    }
+    return devices;
+}
+
+InputCommon::ButtonMapping Adapter::GetButtonMappingForDevice(const Common::ParamPackage& params) {
+    // This list is missing ZL/ZR since those are not considered buttons.
+    // We will add those afterwards
+    // This list also excludes any button that can't be really mapped
+    std::unordered_map<Settings::NativeButton::Values, PadButton> switch_to_gcadapter_button = {
+        {Settings::NativeButton::A, PadButton::PAD_BUTTON_A},
+        {Settings::NativeButton::B, PadButton::PAD_BUTTON_B},
+        {Settings::NativeButton::X, PadButton::PAD_BUTTON_X},
+        {Settings::NativeButton::Y, PadButton::PAD_BUTTON_Y},
+        {Settings::NativeButton::Plus, PadButton::PAD_BUTTON_START},
+        {Settings::NativeButton::DLeft, PadButton::PAD_BUTTON_LEFT},
+        {Settings::NativeButton::DUp, PadButton::PAD_BUTTON_UP},
+        {Settings::NativeButton::DRight, PadButton::PAD_BUTTON_RIGHT},
+        {Settings::NativeButton::DDown, PadButton::PAD_BUTTON_DOWN},
+        {Settings::NativeButton::SL, PadButton::PAD_TRIGGER_L},
+        {Settings::NativeButton::SR, PadButton::PAD_TRIGGER_R},
+        {Settings::NativeButton::R, PadButton::PAD_TRIGGER_Z},
+    };
+    if (!params.Has("port")) {
+        return {};
+    }
+
+    InputCommon::ButtonMapping mapping{};
+    for (const auto& [switch_button, gcadapter_button] : switch_to_gcadapter_button) {
+        Common::ParamPackage button_params({{"engine", "gcpad"}});
+        button_params.Set("port", params.Get("port", 0));
+        button_params.Set("button", static_cast<int>(gcadapter_button));
+        mapping[switch_button] = button_params;
+    }
+
+    // Add the missing bindings for ZL/ZR
+    std::unordered_map<Settings::NativeButton::Values, PadAxes> switch_to_gcadapter_axis = {
+        {Settings::NativeButton::ZL, PadAxes::TriggerLeft},
+        {Settings::NativeButton::ZR, PadAxes::TriggerRight},
+    };
+    for (const auto& [switch_button, gcadapter_axis] : switch_to_gcadapter_axis) {
+        Common::ParamPackage button_params({{"engine", "gcpad"}});
+        button_params.Set("port", params.Get("port", 0));
+        button_params.Set("button", static_cast<int>(PadButton::PAD_STICK));
+        button_params.Set("axis", static_cast<int>(gcadapter_axis));
+        mapping[switch_button] = button_params;
```

Comment:
```suggestion
        mapping,insert_or_assign(switch_button, std::move(button_params));
```

Ditto above for the other loop as well.